### PR TITLE
New handler for exposing state-by-state donation totals.

### PR DESCRIPTION
**Pull request by abliss**
A response to https://mobile.twitter.com/lessig/status/638424623105896449?s=17

### CODE REVIEW

File: backend/handlers.py [hunk 1]
```diff
@@ -722,6 +724,23 @@ def post(self):
...
+# Election Campaign Fund: http://www.fec.gov/press/bkgnd/fund.shtml
+# "... raising at least $5,000 in each of 20 states. Only contributions from
+# individuals, and only contributions up to $250, are matchable"
+class StatesHandler(webapp2.RequestHandler):
+  def get(self):
+    util.EnableCors(self)
+    totals = {}
+    for pledge in model.Pledge.all():
+      user = model.User.all().filter('email =', pledge.email).get()
+      totals[user.state] += min(pledge.amountCents, 25000) # max $250
```
Comment (rickhanlonii) [inline backend/handlers.py:737]: I double checked to make sure that the first $250 counts (opposed to not being allowed to count the whole donation).

Just to be clear, from [here](http://www.fec.gov/press/bkgnd/fund.shtml):

> only $250 of an individual’s contribution can be applied toward the $5,000 threshold in each state.

File: backend/handlers.py [hunk 2]
```diff
@@ -722,6 +724,23 @@ def post(self):
...
+
+    for state, total in totals.iteritems():
+      totals[state] = min(500000, total) # no point counting past $5k
```
Comment (rickhanlonii) [inline backend/handlers.py:740]: IMO you should return the entire total to the client and allow the client to decide if they care about totals over $5k.

**Comment by rickhanlonii** [comment]
Great stuff.

**Comment by aaronlifshin** [comment]
Worried about performance of this.  It needs to be a back end thread that
updates totals every minute or so.  Otherwise if we include it in a webpage
it may be very slow.

A

**Comment by abliss** [comment]
Yeah, I agree it may be slow. I didn't do anything fancier because I'm
having a hard time setting up a viable dev server to test on. I wrote it
this simple way to have best hopes of it working correctly untested.

There's currently no web page pointing at the endpoint, so I was hoping I
could cajole you into pushing it to production (or a test server with a
static pledge dump), then we can test to see just how bad the performance
is.

**Comment by aaronlifshin** [comment]
@furf is going to pull this in and drop it into our test server

File: backend/handlers.py (hunk 2 above)
Comment (abliss) [inline backend/handlers.py:740]: From a technical standpoint I agree. however I speculated (out of thin air) that the exact state totals might be too revealing for the campaign's strategic purposes. Whereas just knowing which states had hit $5k and which ones were close seems no bigger a reveal than what's already gone out to twitter.

Comment (rickhanlonii) [inline backend/handlers.py:740]: Oh, that's a good point.

**Comment by furf** [comment]
Hey @abliss! Thanks for the help. Would you mind creating another pull request to the `feature/state-by-state-totals` branch so that I can run a quick perf check on test before integrating? (GitHub limitation)

**Comment by abliss** [comment]
Sure: https://github.com/Lessig2016/pledgeservice/pull/8

File: backend/handlers.py (hunk 2 above)
Comment (brafkind) [inline backend/handlers.py:740]: Is it actually strategically advantageous to limit the funds total to $5k per state? This campaign also espouses transparency. What would be the harm of showing the actual total?

Comment (aaronlifshin) [inline backend/handlers.py:740]: Showing the total is fine, as long as we are limiting each donation to 250
(qualified amount for federal matching funds)

A

**Pull request closed without merging**